lib/Runtime/Provisioner/Provisioner.cpp

```diff
@@ -105,8 +124,16 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module,
     for (auto &node : device.second) {
       // Only compile if we haven't compiled before. If we have previously
       // compiled the function reuse it.
-      auto it = functions_.find(node->name);
-      if (it == functions_.end()) {
+      bool alreadyCompiled{false};
+      {
+        std::lock_guard<std::mutex> functionLock(functionsLock_);
```

Comment:
It's often more expensive to have a fine grained lock if you take and release it a lot, e.g. in this loop you take and release the lock 3 times per device?

I think you could refactor this to do the compile outside the lock then update all the state under the lock in one go.

Reply:
Good call, I refactored this to pull the lock out of the loop entirely. Now all compiled functions are added to functions_ at the end all in one lock.